### PR TITLE
Support remote app deploy

### PR DESCRIPTION
**Pull request by jshum2479**
This PR adds support for remote updating a domain and deploying applications.  I also refactored some of the sonar complexities warnings in application_deployer.

### CODE REVIEW

File: core/src/main/python/wlsdeploy/tool/util/wlst_helper.py [hunk 1]
```diff
@@ -1361,6 +1361,45 @@ def server_config(self):
...
+            self.server_config()
+            self.cd('/')
+            name = self.get('Name')
+        except self.__load_global('WLSTException'), e:
+            pwe = exception_helper.create_exception(self.__exception_type, 'WLSDPLY-00065',
+                                                    _format_exception(e), error=e)
+            self.__logger.throwing(class_name=self.__class_name, method_name=_method_name, error=pwe)
+            raise pwe
+        return name
+        self.__logger.exiting(class_name=self.__class_name, method_name=_method_name)
```
Comment (ddsharpe) [inline core/src/main/python/wlsdeploy/tool/util/wlst_helper.py:1381]: Should this logging call be before the return?

**Comment by ddsharpe** [comment]
Please double check your copyrights.  core/src/main/python/wlsdeploy/tool/deploy/applications_deployer.py needs to be updated to 2023, not sure about others

**Review by robertpatrick — CHANGES_REQUESTED** [review]
Please remove WDT 3.0 doc changes from the PR and put into the branch for PR #1340

File: core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java [hunk 2]
```diff
@@ -495,6 +495,52 @@ public boolean containsFileOrPath(String path) throws WLSDeployArchiveIOExceptio
...
+        return result;
+    }
+
+    /** Check whether the path should be used for remote update
+     *
+     * @param path
+     * @return true if the path should not be used for remote update
+     * @throws IllegalArgumentException    if the path is null or empty
+     */
+    public boolean isRemoteUpdateDomainForbiddenPath(String path)  throws IllegalArgumentException  {
```
Comment (robertpatrick) [inline core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java:528]: I'm not sure that this is the best place for this logic.  In my opinion, the archive file shouldn't know anything about remote mode.  Can't we put this into archive_helper.py?

File: core/src/main/python/deploy.py [hunk 3]
```diff
@@ -41,12 +41,12 @@
 
 __required_arguments = [
     CommandLineArgUtil.ORACLE_HOME_SWITCH,
-    CommandLineArgUtil.DOMAIN_HOME_SWITCH,
     CommandLineArgUtil.MODEL_FILE_SWITCH
 ]
 
 __optional_arguments = [
     # Used by shell script to locate WLST
+    CommandLineArgUtil.DOMAIN_HOME_SWITCH,
```
Comment (robertpatrick) [inline core/src/main/python/deploy.py:49]: Is there ever a case where the model for a remote deploy operation might have a domain home token?  If so, what happens if the user did not provide the domain home path on the remote machine?

File: core/src/main/python/update.py [hunk 4]
```diff
@@ -44,12 +46,12 @@
 
 __required_arguments = [
     CommandLineArgUtil.ORACLE_HOME_SWITCH,
-    CommandLineArgUtil.DOMAIN_HOME_SWITCH,
```
Comment (robertpatrick) [inline core/src/main/python/update.py:47]: Same questions as for Deploy.  For discoverDomain -remote, we still require the user to provide -domain_home <remote-domain-home-path> to support tokenization.  I would think we might need this for at least updateDomain -remote if not all remote operations.

**Review by robertpatrick — COMMENTED** [review]
Comments inline

File: core/src/main/python/wlsdeploy/tool/util/wlst_helper.py (hunk 1 above)
Comment (jshum2479) [reply core/src/main/python/wlsdeploy/tool/util/wlst_helper.py:1380]: Fixed

**Review by robertpatrick — APPROVED** [review]

**Pull request merged**